### PR TITLE
refactor pam-u2f: add keysPath, verbose, fix docs about u2f_keys path

### PR DESCRIPTION
**Pull request by spinus**
- change enableU2F option to U2F.\* set
- added few pam-u2f options
- fix documentation about default u2f_keys locations

### CODE REVIEW

File: nixos/modules/security/pam.nix [hunk 1]
```diff
@@ -36,14 +36,61 @@ let
...
 
-      u2fAuth = mkOption {
-        default = config.security.pam.enableU2F;
-        type = types.bool;
-        description = ''
-          If set, users listed in
-          <filename>~/.yubico/u2f_keys</filename> are able to log in
-          with the associated U2F key.
-        '';
+      U2F = {
```
Comment (copumpkin) [inline nixos/modules/security/pam.nix]: Not a fan of starting the module name with uppercase, but am not sure if we have any official guidelines on that.

File: nixos/modules/security/pam.nix [hunk 2]
```diff
@@ -36,14 +36,61 @@ let
...
+        enabled = mkOption {
+          default = config.security.pam.U2F.enabled;
```
Comment (copumpkin) [inline nixos/modules/security/pam.nix]: Isn't this a circular default? Or what am I missing?

Comment (spinus) [inline nixos/modules/security/pam.nix]: :) yup.

**Comment by copumpkin** [comment]
Why the duplicate module?

File: nixos/modules/security/pam.nix (hunk 1 above)
Comment (spinus) [inline nixos/modules/security/pam.nix]: I'm big fan of "this-notation" or "this_notation" but I'm not sure we want that here. Some packages are using this convention, but I didn't see that for options.

Here I followed something like, camelCase but if the first letter is big I'm making it smaller just for naming sake. the same I would do if the name would started with API I wouldn't do apiSomething.

But I can use whatever convention is agreed :-)

Comment (copumpkin) [inline nixos/modules/security/pam.nix]: https://github.com/Yubico/pam-u2f Yubico themselves seem to have no trouble keeping it lowercase

Comment (spinus) [inline nixos/modules/security/pam.nix]: True.
Is that good occasion to move that decision to mailing list or ticket and make some naming convention agreements? I saw one ticket about that but as far as I remember it stucked.

**Comment by spinus** [comment]
What do you mean?

**Comment by copumpkin** [comment]
@spinus as far as I can tell, you have two copies of all the U2F options, first in the `let` block, and then in the final returned attrset. They have identical (as far as I can eyeball) documentation, types, and names.

**Comment by spinus** [comment]
The reason for this is (I guess), the "top" layer works as defaults for all pam services, the other one is per service.

example:
```
config.security.pam.U2F.enabled = True
config.security.pam.services.<name?>.U2F.enabled=False
```
So you can enable it globally and disable for particular modules.
For each service there is the same set of options which get defaults from "global" services config.
It was like that before but only one option was passed down the hole and the name was different. My idea is to unify this names to understand that you are changing the same thing (previously the names were enabledU2F and u2fAuth, so not only different name but different naming convention as well).

The problem I have here, I have no idea how to grab description from "outer" scope the same I grabbed the default value so I copy pasted as "good" start :)

**Comment by spinus** [comment]
hey, can we move it a little? What I should do/fix to make this happen?

**Comment by spinus** [comment]
@copumpkin hey, updated the name (U2F -> u2f), is it better now?

**Comment by spinus** [comment]
@philandstuff @edolstra @obadz do you think this is good direction? Any idea how I could use description only once in current situation (I know I could extract it level up as variable, but that would be confusing probably)

**Comment by jagajaga** [comment]
ping @copumpkin

**Comment by spacekitteh** [comment]
@grahamc security :)

File: nixos/modules/security/pam.nix [hunk 3]
```diff
@@ -36,14 +36,61 @@ let
...
-      u2fAuth = mkOption {
-        default = config.security.pam.enableU2F;
-        type = types.bool;
-        description = ''
-          If set, users listed in
-          <filename>~/.yubico/u2f_keys</filename> are able to log in
-          with the associated U2F key.
-        '';
+      u2f = {
+        enabled = mkOption {
```
Comment (fpletz) [inline nixos/modules/security/pam.nix]: The attribute shoud be called `enable` like in every other module.

File: nixos/modules/security/pam.nix [hunk 4]
```diff
@@ -412,11 +459,59 @@ in
...
+        type = types.bool;
+        description = ''
+          By default <literal>pam-u2f</literal> module does not inform user
+          that he needs to use the u2f device, it just waits without a prompt.
+
+          If you sen this option to <literal>true</literal>,
+          <literal>cue</literal> option is added to <literal>pam-u2f</literal>
+          module and reminder message will be displayed.
+        '';
+      };
```
Comment (fpletz) [inline nixos/modules/security/pam.nix]: Both `u2f` options should be generated by a function instead auf copy-paste.

File: nixos/modules/security/pam.nix [hunk 5]
```diff
@@ -36,14 +36,61 @@ let
...
+          default = config.security.pam.u2f.verbose;
+          type = types.bool;
+          description = ''
+            By default <literal>pam-u2f</literal> module does not inform user
+            that he needs to use the u2f device, it just waits without a prompt.
+
+            If you sen this option to <literal>true</literal>,
+               <literal>cue</literal> option is added to <literal>pam-u2f</literal>
+                 module and reminder message will be displayed.
+                 '';
```
Comment (fpletz) [inline nixos/modules/security/pam.nix]: Wrong indention.

**Review by fpletz — CHANGES_REQUESTED** [review]

File: nixos/modules/security/pam.nix (hunk 4 above)
Comment (spinus) [inline nixos/modules/security/pam.nix]: @fpletz now, one year after I pushed that, I know lot more about nix :-) I try to refactor.

Thank you for your review.

**Comment by jb55** [comment]
status of this?

**Comment by spinus** [comment]
@jb55 waits for me to refactor it, my current estimation is I'll take a look in next 2 weeks.

**Comment by makefu** [comment]
any update on this? would love to see this in master 👍

**Comment by spinus** [comment]
Yeah, I started working on it again in January, but during that year stuff has changed a little and I started doing it again rather then rebasing or anything. In the meantime I stuck somewhere there (I had a problem with nix expression evaluation). 

I was also wondering if the name should be changed to yubico (from u2f). Yubico has pam module named u2f but it looks like yubikey is supported but I'm not sure if that is universal or not. If it's universal, it's kine of unfortunate that default key files are named yubikey. What do you think?

**Comment by jb55** [comment]
Tomasz Czyż <notifications@github.com> writes:

> I was also wondering if the name should be changed to yubico (from u2f). Yubico has pam module named u2f but it looks like yubikey is supported but I'm not sure if that is universal or not. If it's universal, it's kine of unfortunate that default key files are named yubikey. What do you think?

It is universal, I have a trezor which works with pam-u2f.

**Comment by spinus** [comment]
@jb55 thanks for info. In that case I'll just continue to refactor it as u2f. If I won't finish it to the end of mid-April, I'll post what I have until then and maybe somebody can continue.

**Comment by spinus** [comment]
btw, I was thinking about writing a test for that module but I have no idea how to test that in the vm without the key :-) Any idea how we can test that module?

**Comment by spinus** [comment]
OK, I understood a little better the structure in pam.nix and tried to extend it rather than changing convention like in my previous PR. So security.pam.u2f defines the options and they are set for all applications that use pam. Particular pam applications can use of not u2f by setting u2fAuth attribute (looks like that's how all other modules are set).

I added few (but not all) pam-u2f options.
I added really bad test, probably would be nice to have "software" device and do proper test like in oath module, but I'm not that familiar with u2f and don't have time to dive into it now. If anyone could provide any input I'm happy to implement proper test.

Please review :)

**Comment by jb55** [comment]
Tomasz Czyż <notifications@github.com> writes:
> Please review :)

I got around to testing this today. It seems to work for the initial
login and password prompts, but for some reason it breaks systemd user
services...

**Comment by spinus** [comment]
@jb55 could you describe a little more the use case? I'll try to test that and fix.

**Comment by jb55** [comment]
Sure, if I remember correctly all I did was set it to `required` and
tried logging in. That worked fine but I noticed redshift wasn't
running. So I tried `systemctl start --user redshift` and I got an
error. I didn't have time to look into it further.

We could merge this and open a separate issue, as it seemed to work fine
other than that small hiccup.

**Comment by spinus** [comment]
ah no, I would like not to break anything what is working, give me some time, I'll try to test it or even write test case or try to look if it's possible to limit u2f module to affect only "human" users (if it's affecting system users or whatever)

**Comment by ixmatus** [comment]
I wanted to give my encouragement for this PR and that I'm willing to help if needed and where my spare time permits.

**Comment by spinus** [comment]
@fpletz please take a look

**Comment by spinus** [comment]
@fpletz please take a look

File: nixos/modules/security/pam.nix [hunk 6]
```diff
@@ -37,12 +37,11 @@ let
       };
 
       u2fAuth = mkOption {
-        default = config.security.pam.enableU2F;
+        default = config.security.pam.u2f.enable;
```
Comment (kalbasit) [inline nixos/modules/security/pam.nix]: can you please update https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/rename.nix to reflect this rename?

File: nixos/modules/security/pam.nix [hunk 7]
```diff
@@ -450,11 +449,86 @@ in
...
+
+          File format is:
+          <literal>username:first_keyHandle,first_public_key: second_keyHandle,second_public_key</literal>
+          This file can be generated using <command>pamu2fcfg</command> command.
+
+          More about that you can read on
+          <literal>https://developers.yubico.com/pam-u2f/</literal>.
+        '';
+      };
+      authfile = mkOption {
```
Comment (kalbasit) [inline nixos/modules/security/pam.nix]: can you add a new line between options? See above (same file) for examples.

**Review by kalbasit — CHANGES_REQUESTED** [review]

File: nixos/modules/security/pam.nix [hunk 8]
```diff
@@ -450,11 +449,86 @@ in
...
+          File format is:
+          <literal>username:first_keyHandle,first_public_key: second_keyHandle,second_public_key</literal>
+          This file can be generated using <command>pamu2fcfg</command> command.
+
+          More about that you can read on
+          <literal>https://developers.yubico.com/pam-u2f/</literal>.
+        '';
+      };
+      control = mkOption {
+        default = "required";
```
Comment (kalbasit) [inline nixos/modules/security/pam.nix]: It's currently set to `sufficient`, setting it to `required` might unexpectedly change the behavior for everyone. Please set this back to `sufficient`

**Review by kalbasit — CHANGES_REQUESTED** [review]

File: nixos/tests/pam-u2f.nix [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+  machine =
+    { config, pkgs, lib, ... }:
+    {
+      security.pam.u2f = {
+        enable = true;
+        interactive=true;
+        control="sufficient";
+        cue=true;
+        debug=true;
```
Comment (kalbasit) [inline nixos/tests/pam-u2f.nix]: please leave a space around the `=` signs

File: nixos/tests/pam-u2f.nix [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+      };
+    };
+
+  testScript =
+    ''
+      $machine->waitForUnit('multi-user.target');
+      $machine->succeed('egrep "auth sufficient .*u2f.*" /etc/pam.d/ -R ');
+
```
Comment (kalbasit) [inline nixos/tests/pam-u2f.nix]: nitpick: remove newline.

**Review by kalbasit — CHANGES_REQUESTED** [review]

**Pull request closed without merging**